modules/apigee_edge_teams/src/TeamPermissionHandler.php

```diff
@@ -0,0 +1,300 @@
+ *   # (optional) Additional description for the team permission used in the UI.
+ *   description: 'Add/remove team members.'
+ *   # (optional) The category that the team permission belongs (ex.: Team Apps)
+ *   # to be shown on the team permission administration page.
+ *   # Default is the name of the provider module.
+ *   category: 'Members'
+ *
+ * # An array of classes used to generate dynamic team permissions.
+ * permission_providers:
+ *   # Each item in the array must implement
+ *   # DynamicTeamPermissionProviderInterface interface.
+ *   - Drupal\apigee_edge_teams\DefaultTeamPermissionsProvider
+ * @endcode
+ *
+ * Based on Drupal core's PermissionHandler with a bunch of improvements.
+ *
+ * @see \Drupal\user\PermissionHandler
+ * @see \Drupal\apigee_edge_teams\DynamicTeamPermissionProviderInterface
+ */
+final class TeamPermissionHandler implements TeamPermissionHandlerInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * The module handler.
+   *
+   * @var \Drupal\Core\Extension\ModuleHandlerInterface
+   */
+  private $moduleHandler;
+
+  /**
+   * The class resolver.
+   *
+   * @var \Drupal\Core\DependencyInjection\ClassResolverInterface
+   */
+  private $classResolver;
+
+  /**
+   * The YAML discovery object instance cache.
+   *
+   * Yaml discovery to find all .team_permissions.yml files.
+   *
+   * Use getYamlDiscovery() instead.
+   *
+   * @var \Drupal\Core\Discovery\YamlDiscovery|null
+   */
+  private $yamlDiscovery;
+
+  /**
+   * The team member role entity storage.
+   *
+   * @var \Drupal\apigee_edge_teams\Entity\Storage\TeamMemberRoleStorageInterface
+   */
+  private $teamMemberRoleStorage;
+
+  /**
+   * The team membership manager service.
+   *
+   * @var \Drupal\apigee_edge_teams\TeamMembershipManagerInterface
+   */
+  private $teamMembershipManager;
+
+  /**
+   * The team role storage.
+   *
+   * @var \Drupal\apigee_edge_teams\Entity\Storage\TeamRoleStorageInterface
+   */
+  private $teamRoleStorage;
+
+  /**
+   * TeamPermissionHandler constructor.
+   *
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler.
+   * @param \Drupal\Core\DependencyInjection\ClassResolverInterface $class_resolver
+   *   The class resolver.
+   * @param \Drupal\apigee_edge_teams\TeamMembershipManagerInterface $team_membership_manager
+   *   The team membership manager service.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   */
+  public function __construct(ModuleHandlerInterface $module_handler, ClassResolverInterface $class_resolver, TeamMembershipManagerInterface $team_membership_manager, EntityTypeManagerInterface $entity_type_manager) {
+    $this->moduleHandler = $module_handler;
+    $this->classResolver = $class_resolver;
+    $this->teamMembershipManager = $team_membership_manager;
+    $this->teamRoleStorage = $entity_type_manager->getStorage('team_role');
```

Comment:
I marked the wrong file for this issue but see: https://github.com/apigee/apigee-edge-drupal/pull/138#discussion_r260523610